File: 1431-all-ancestors-of-a-node-in-a-directed-acyclic-graph/1431-all-ancestors-of-a-node-in-a-directed-acyclic-graph.cpp
#pragma GCC optimize(O3, "unroll-loops");

class Solution {
private:
    vector<vector<int>> p;
    vector<vector<int>> g;
    void findParDFS(int prev, int curr) {
        for (int adj: g[curr]) {
            if (p[adj].empty() || p[adj].back() != prev) {
                p[adj].push_back(prev);
                findParDFS(prev, adj);
            }
        }
    }
public:
    vector<vector<int>> getAncestors(int n, vector<vector<int>>& edges) {
        p = vector<vector<int>>(n, vector<int>());
        g = vector<vector<int>>(n, vector<int>());
        for (int i = 0; i < edges.size(); i++) g[edges[i][0]].push_back(edges[i][1]);
        for (int i = 0; i < n; i++) findParDFS(i, i);
        return p;
    }
};

auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();

File: 1431-all-ancestors-of-a-node-in-a-directed-acyclic-graph/1431-all-ancestors-of-a-node-in-a-directed-acyclic-graph.java
class Solution {
    private List<Integer>[] g;
    private List<List<Integer>> p;
    private void findParDFS(int prev, int curr) {
        for (int adj: g[curr]) {
            if (p.get(adj).isEmpty() || p.get(adj).lastIndexOf(prev) != p.get(adj).size() - 1) {
                p.get(adj).add(prev);
                findParDFS(prev, adj);
            }
        }
    }
    public List<List<Integer>> getAncestors(int n, int[][] edges) {
        g = new ArrayList[n];
        p = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            g[i] = new ArrayList<>();
            p.add(new ArrayList<>());
        }

        for (int i = 0; i < edges.length; i++) g[edges[i][0]].add(edges[i][1]);
        for (int i = 0; i < n; i++) findParDFS(i, i);

        return p;
    }
}